packages/repo/src/sync.ts

```diff
@@ -1,33 +1,118 @@
-import * as auth from '@atproto/auth'
-import { CID } from 'multiformats/cid'
-import { RepoStorage } from './storage'
-import { DataDiff } from './mst'
+import { MemoryBlockstore, RepoStorage } from './storage'
+import DataDiff from './data-diff'
 import Repo from './repo'
 import * as verify from './verify'
+import { DidResolver } from '@atproto/did-resolver'
+import * as util from './util'
+import { CommitData, RecordWriteOp, RepoContents } from './types'
+import { CID } from 'multiformats/cid'
+import CidSet from './cid-set'
+import { MissingBlocksError } from './error'
 
-export const loadRepoFromCar = async (
-  carBytes: Uint8Array,
+export const loadCheckout = async (
   storage: RepoStorage,
-  verifier: auth.Verifier,
-): Promise<Repo> => {
-  const { root } = await storage.loadDiff(carBytes, (root: CID) => {
-    return verify.verifyUpdates(storage, null, root, verifier)
-  })
-  return Repo.load(storage, root)
+  repoCar: Uint8Array,
+  didResolver: DidResolver,
+): Promise<{ root: CID; contents: RepoContents }> => {
+  const { root, blocks } = await util.readCar(repoCar)
+  const updateStorage = new MemoryBlockstore(blocks)
+  const checkout = await verify.verifyCheckout(updateStorage, root, didResolver)
+
+  const checkoutBlocks = await updateStorage.getBlocks(
+    checkout.newCids.toList(),
+  )
+  if (checkoutBlocks.missing.length > 0) {
+    throw new MissingBlocksError('sync', checkoutBlocks.missing)
+  }
+  await Promise.all([
+    storage.putMany(checkoutBlocks.blocks),
+    storage.updateHead(root, null),
+  ])
+
+  return {
+    root,
+    contents: checkout.contents,
+  }
+}
+
+export const loadFullRepo = async (
+  storage: RepoStorage,
+  repoCar: Uint8Array,
+  didResolver: DidResolver,
+): Promise<{ root: CID; ops: RecordWriteOp[] }> => {
```

Comment:
Nothing to do about this here/now I don't think, but just a thought rattling around my brain.  If someone is syncing these `ops` into e.g. their own index, and they want to keep their index consistent, they'd need to process all `ops` at once transactionally.  If the `ops` were broken-up by commit, then they could gradually process commit-by-commit (or condensed into batches of commits), landing at a consistent state each step along the way.  It would be useful to be able to ask something like "give me this history in ops batches of size roughly 100, always containing full commits."

Reply:
That's a good point. My motivation here was actually to simplify processing ops: for instance, if someone creates & deletes a record several times over a commit range, this will collapse it into one operation.

On second look, this should probably be an optional processing step done by the consumer: we return ops per commit & they can process them atomically by commit, or they can collapse them down into one big set of ops

Reply:
this is actually trickier than i thought it'd be. you lose some info when going from diff -> op. specifically the CID associated with the del. It may make sense to start tracking cids in the operations as well as the records & include the cid in the delete operation as well 🤔

Reply:
Yeah hmmm, I can see that.  Could be nice!  If it's too much of a rabbit hole for this PR though, I'm certainly not too hung-up on it right now.

Reply:
Alright I ended up doing it, because it actually does make these functions much nicer 

Split it out into it's own PR here: https://github.com/bluesky-social/atproto/pull/460